pkgs/development/compilers/gcc/common/configure-flags.nix

```
@@ -111,9 +111,30 @@ let
       "--with-mpc=${libmpc}"
     ]
     ++ lib.optional (libelf != null) "--with-libelf=${libelf}"
-    ++ lib.optional (!(crossMingw && crossStageStatic))
+    ++ lib.optionals (!(crossMingw && crossStageStatic)) [
       "--with-native-system-header-dir=${lib.getDev stdenv.cc.libc}/include"
 
+      # gcc builds for cross-compilers (build != host) or cross-built
+      # gcc (host != target) always apply the offset prefix to disentangle
+      # target headers from build or host headers:
+      #     ${with_build_sysroot}${native_system_header_dir}
+      #  or ${test_exec_prefix}/${target_noncanonical}/sys-include
+      #  or ${with_sysroot}${native_system_header_dir}
+      # While native build (build == host == target) uses passed headers
+      # path as is:
+      #    ${native_system_header_dir}
+      #
+      # Nixpkgs uses flat directory structure for both native and cross
+      # cases. As a result libc headers don't get found for cross case
+      # and many modern features get disabled (libssp is used instead of
+      # target-specific implementations and similar). More details at:
+      #   https://github.com/NixOS/nixpkgs/pull/181802#issuecomment-1186822355
+      #
+      # We pick "/" path to effectively avoid sysroot offset and make it work
+      # as a native case.
+      "--with-build-sysroot=/"
```

probably should do the following but env vars can't be used here
```suggestion
      "--with-build-sysroot=$NIX_BUILD_TOP"
```

---

I don't think so. `$NIX_BUILD_TOP` is a temporary build directory. For example:

```
$ nix develop nixpkgs#hello
$$ echo $NIX_BUILD_TOP
/tmp/nix-shell.fofHiz
```

The goal here is not to prepend any real prefix and rely on absolute paths. I'm not sure `$NIX_BUILD_TOP` is ever useful for external paths.